### ISSUE TITLE
Leveraging std:: machinery

### ISSUE THREAD

**Issue by elfprince13** [GarageGames/Torque3D#1964]
There are a variety of subsystems within Torque that I think would benefit from using the standard library more. While there would be some overhead here in terms of implementing the changes, I believe in the long run they would drastically reduce the maintenance costs for the engine, as well as the potential for bugs in various subtle implementation issues.

The example that's currently sticking out to me is the huge amount of manual reference counting machinery that could be elided through appropriate use of `shared_ptr` (and probably some edge cases where using reference types, `weak_ptr`, and `unique_ptr` might be more appropriate entirely).

Similarly, keeping around (and maintaining) custom implementations of `vector`, `map`, etc, seems like a poor use of development investment, unless there are specific performance reasons for implementing our own.

**Comment by Areloch** [comment]
On the vector/map side of things, there's definitely been some investigation/discussion about overhead on the std libraries. The highlights I remember for vector at least is that in debug, it has a bunch of extra error checking/handling logic and can get comparitively rather slow(though not an issue in release). I think it has has some curious memory handling decisions when incrementing vs preallocating, but I don't remember the particulars of that.

That said, now is a good time for us to actually draft some benchmarks to get a definitive picture of the performance for these sorts of things, and if it makes sense, look to shifting over.

For the *_ptr stuff, also worth looking at. Presumably that would largely replace the refbase stuff we utilize now, with the weak references and all that?

**Comment by elfprince13** [comment]
Comparative slowness in debug builds for better error handling logic sounds like a good thing to me, but I guess this is a good discussion to have now. It's worth noting if people are concerned about specifics around allocation that one option is to provide a partial specialization in the Torque namespace, with whatever custom allocation stuff we want.

I believe you're correct w.r.t. to the refbase stuff.

**Comment by Areloch** [comment]
Yeah, actual numbers would be key, but I think that having the extra assurance in debug wouldn't be a bad thing unless it's horrifyingly slow for some oddball reason.

Same deal for the allocations for vector and stuff. If the numbers show it's a relatively minor perf cost, it may not matter enough to fret. If it's a pretty sizeable one, then as you say some specialist handling in such cases would be a good compromise.

**Comment by JeffProgrammer** [comment]
I tend to avoid std::shared_ptr because of the baggage and overhead it brings. It has to guarantee thread safety even if the pointer never touches another thread, because it uses atomic reference counting.

std::unique_ptr is fine, same as new/delete without additional overhead.

I know, I know, premature optimization at best, but still. Using unnecessary atomic reference counting in code that could be used a lot if something to be careful of...

**Comment by dottools** [comment]
Atomic operations doesn't mean it'll be slow, pretty much all integer based atomic operations are inexpensive time wise. Worst case causes a cache miss up to L2 or L3 memory on multiple core systems which is nothing compared to a cache miss to system memory.
